types/+MPCT/code_MPCT_ADMM_semiband_C.c

```diff
@@ -0,0 +1,930 @@
+                    }
+                    else{
+                        aux_1[i] -= A[i-l*nn_][j] * xi[(l-1)*nm_+j];
+                    }
+
+                }
+
+                for (unsigned int j = 0 ; j < mm_ ; j++){
+                    aux_1[i] -= B[i-l*nn_][j] * xi[(l-1)*nm_+nn_+j];
+                }
+
+            }
+
+        }
+
+        for (unsigned int i = (NN_+1)*nn_ ; i < (NN_+2)*nn_ ; i++){
+
+            for (unsigned int j = 0 ; j < nn_ ; j++){
+
+                if (i-(NN_+1)*nn_ == j){
+                    aux_1[i] -= (A[i-(NN_+1)*nn_][j]-1.0) * xi[NN_*nm_+j];
+                }
+                else{
+                    aux_1[i] -= A[i-(NN_+1)*nn_][j] * xi[NN_*nm_+j];
+                }
+
+            }
+
+            
+            for (unsigned int j = 0 ; j < mm_ ; j++){
+                aux_1[i] -= B[i-(NN_+1)*nn_][j] * xi[NN_*nm_+nn_+j];
+            }
+
+        }
+        // End of computation of -(G*xi+b)
+
+        solve_banded_Chol(Alpha, Beta, z1_b, aux_1); // Obtains z1_b
+
+        memset(z2, 0, sizeof(double)*2*nm_);
+
+        // Computation of z2_b
+        for  (unsigned int i = 0 ; i < 2*nm_ ; i++){ 
+
+            for(unsigned int j = 0 ; j < nn_ ; j++){
+
+                z2[i] += M_tilde[i][j] * z1_b[j]; // M_tilde is dense, but it presents repetitions inside, so we use a shortened version of it instead
+
+            }
+
+            for (unsigned int l = 0 ; l < NN_-1 ; l++){
+            
+                for (unsigned int j = nn_ ; j < 2*nn_ ; j++){
+                
+                    z2[i] += M_tilde[i][j] * z1_b[l*nn_+j];
+
+                }
+                
+            }
+
+            for (unsigned int j = 2*nn_ ; j < 4*nn_ ; j++){
+
+                z2[i] += M_tilde[i][j] * z1_b[(NN_-2)*nn_+j];
+
+            }
+
+        }
+        // End of computation of z2_b
+
+        memset(aux_1, 0, sizeof(double)*(NN_+2)*nn_);
+
+        // Computation of (U_tilde*z2_b)
+        for  (unsigned int i = 0 ; i < 2*nm_ ; i++){ 
+
+            for(unsigned int j = 0 ; j < nn_ ; j++){
+
+                aux_1[j] += U_tilde[j][i] * z2[i]; // U_tilde is dense, but it presents repetitions inside, so we use a shortened version of it instead
+
+            }
+
+            for (unsigned int l = 0 ; l < NN_-1 ; l++){
+            
+                for (unsigned int j = nn_ ; j < 2*nn_ ; j++){
+                
+                    aux_1[l*nn_+j] += U_tilde[j][i] * z2[i];
+
+                }
+                
+            }
+
+            for (unsigned int j = 2*nn_ ; j < 4*nn_ ; j++){
+
+                aux_1[(NN_-2)*nn_+j] += U_tilde[j][i] * z2[i];
+
+            }
+
+        }
+        // End of computation of (U_tilde*z2_b)
+
+        solve_banded_Chol(Alpha, Beta, z3_b, aux_1); // Obtains z3_b
+
+        // Computation of mu, which is the solution of eq. (9b)
+        for (unsigned int i = 0 ; i < (NN_+2)*nn_ ; i++){ 
+
+            mu[i] = z1_b[i] - z3_b[i];
```

Comment:
I think `z1_b` could also be eliminated by using `mu` where `z1_b` is currently used. The code would be more confusing, but we could save some memory. This line here would be `mu[i] -= z3_b[i]`.